### PR TITLE
[IOTDB-2445]select count(*) execution failed : overlapped data should be consumed first

### PR DESCRIPTION
**Pull request by choubenson**
(1) Problem: Currently, cross space compaction may cause unseq data in target files.
(2) Reason: When selecting the source files, it will skip files which are being compacted. For example, seq files 1 2 3 have overlap with unseq file 4, but the file 1 2 are being compacted, so they will only choose file 3 and 4 as the source files in this cross space compaction, this will lead a overlap between the target file and seq files 1 2.
(3) Solution: When selecting files, we need to select all seq files that overlap with unseq files (no matter whether they are being compacted or not). After selecting files, we need to determine whether there are source files being compacted before throwing the task into the queue. If there are, this compaction task is discarded.

As a conclusion, this pr fix cross space compaction selector bug to avoid unseq data in target files.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java [hunk 1]
```diff
@@ -64,24 +64,37 @@
...
-    this.seqFiles = seqFiles.stream().filter(this::filterResource).collect(Collectors.toList());
-    this.unseqFiles = unseqFiles.stream().filter(this::filterResource).collect(Collectors.toList());
+    this.seqFiles = seqFiles.stream().filter(this::filterSeqResource).collect(Collectors.toList());
+    filterUnseqResource(unseqFiles);
   }
 
   /** If returns true, it means to participate in the merge */
-  private boolean filterResource(TsFileResource res) {
+  private boolean filterSeqResource(TsFileResource res) {
     return res.getTsFile().exists()
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java:73]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java [hunk 2]
```diff
@@ -64,24 +64,37 @@
...
+        && (!res.isClosed() || res.stillLives(ttlLowerBound));
+  }
+
+  /** Filter the unseq files into the compaction */
+  private void filterUnseqResource(List<TsFileResource> unseqResources) {
+    for (TsFileResource resource : unseqResources) {
+      if (resource.isCompacting() || resource.isCompactionCandidate() || !resource.isClosed()) {
+        if (unseqFiles.size() > 0) {
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java:84]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/RewriteCrossSpaceCompactionSelector.java [hunk 3]
```diff
@@ -86,9 +88,6 @@ public void selectAndSubmit() {
     if (seqFileList.isEmpty() || unSeqFileList.isEmpty()) {
       return;
     }
-    if (unSeqFileList.size() > config.getMaxCompactionCandidateFileNum()) {
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/RewriteCrossSpaceCompactionSelector.java:89]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java [hunk 4]
```diff
@@ -64,24 +64,37 @@
   private boolean cacheDeviceMeta = false;
 
   public CrossSpaceMergeResource(List<TsFileResource> seqFiles, List<TsFileResource> unseqFiles) {
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java:66]: later in the future :）

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 5]
```diff
@@ -354,7 +354,7 @@
   private int avgSeriesPointNumberThreshold = 10000;
 
   /** Only compact the sequence files */
-  private boolean enableSeqSpaceCompaction = true;
+  private boolean enableSeqSpaceCompaction = false;
 
   /** Only compact the unsequence files */
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java]: resolved

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 6]
```diff
@@ -354,7 +354,7 @@
   private int avgSeriesPointNumberThreshold = 10000;
 
   /** Only compact the sequence files */
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java [hunk 7]
```diff
@@ -64,24 +64,37 @@
...
-    this.seqFiles = seqFiles.stream().filter(this::filterResource).collect(Collectors.toList());
-    this.unseqFiles = unseqFiles.stream().filter(this::filterResource).collect(Collectors.toList());
+    this.seqFiles = seqFiles.stream().filter(this::filterSeqResource).collect(Collectors.toList());
+    filterUnseqResource(unseqFiles);
   }
 
   /** If returns true, it means to participate in the merge */
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java:71]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java [hunk 8]
```diff
@@ -64,24 +64,37 @@
...
+  private boolean filterSeqResource(TsFileResource res) {
     return res.getTsFile().exists()
         && !res.isDeleted()
-        && (!res.isClosed() || res.stillLives(ttlLowerBound))
-        && !res.isCompacting()
-        && !res.isCompactionCandidate();
+        && (!res.isClosed() || res.stillLives(ttlLowerBound));
+  }
+
+  /** Filter the unseq files into the compaction */
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/manage/CrossSpaceMergeResource.java:78]: resolved

File: server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java [hunk 9]
```diff
@@ -251,12 +252,19 @@ private void selectOverlappedSeqFiles(TsFileResource unseqFile) {
...
+        if (!seqFile.mayContainsDevice(deviceId)) {
           continue;
         }
-        // the open file's endTime is Long.MIN_VALUE, this will make the file be filtered below
-        long seqEndTime = seqFile.isClosed() ? seqFile.getEndTime(deviceId) : Long.MAX_VALUE;
-        if (unseqEndTime <= seqEndTime) {
+
+        long seqEndTime = seqFile.getEndTime(deviceId);
+        long seqStartTime = seqFile.getStartTime(deviceId);
+        if (unseqEndTime < seqStartTime) {
```
[the remark this answers is not in the document]
Comment (choubenson) [reply server/src/main/java/org/apache/iotdb/db/engine/compaction/cross/rewrite/selector/RewriteCompactionFileSelector.java:261]: resolved

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**